### ISSUE TITLE
Compute Shader Example Doesn't Seem To Work

### ISSUE THREAD

**Issue by TheCleric** [godotengine/godot#85874]
**Your Godot version:**

v4.2.stable.official [46dc27791]

**Issue description:**

When running a scene with the compute shader example code from the documentation (https://docs.godotengine.org/en/stable/tutorials/shaders/compute_shaders.html), the output does not seem to match what would be expected.

I've added the following to a file in my project called `compute_example.glsl`:

```glsl
#[compute]
#version 450

// Invocations in the (x, y, z) dimension
layout(local_size_x = 2, local_size_y = 1, local_size_z = 1) in;

// A binding to the buffer we create in our script
layout(set = 0, binding = 0, std430) restrict buffer MyDataBuffer {
	float data[];
}
my_data_buffer;

// The code we want to execute in each invocation
void main() {
	// gl_GlobalInvocationID.x uniquely identifies this invocation across all work groups
	my_data_buffer.data[gl_GlobalInvocationID.x] *= 2.0;
}
```

And the following script to my scene which is just a Node2D object:

```GDScript
extends Node2D

# Called when the node enters the scene tree for the first time.
func _ready():
	# Create a local rendering device.
	var rd := RenderingServer.create_local_rendering_device()

	# Load GLSL shader
	var shader_file := load("res://compute_example.glsl")
	var shader_spirv: RDShaderSPIRV = shader_file.get_spirv()
	var shader := rd.shader_create_from_spirv(shader_spirv)

	# Prepare our data. We use floats in the shader, so we need 32 bit.
	var input := PackedFloat32Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
	var input_bytes := input.to_byte_array()

	# Create a storage buffer that can hold our float values.
	# Each float has 4 bytes (32 bit) so 10 x 4 = 40 bytes
	var buffer := rd.storage_buffer_create(input_bytes.size(), input_bytes)

	# Create a uniform to assign the buffer to the rendering device
	var uniform := RDUniform.new()
	uniform.uniform_type = RenderingDevice.UNIFORM_TYPE_STORAGE_BUFFER
	uniform.binding = 0 # this needs to match the "binding" in our shader file
	uniform.add_id(buffer)
	var uniform_set := rd.uniform_set_create([uniform], shader, 0) # the last parameter (the 0) needs to match the "set" in our shader file

	# Create a compute pipeline
	var pipeline := rd.compute_pipeline_create(shader)
	var compute_list := rd.compute_list_begin()
	rd.compute_list_bind_compute_pipeline(compute_list, pipeline)
	rd.compute_list_bind_uniform_set(compute_list, uniform_set, 0)
	rd.compute_list_dispatch(compute_list, 5, 1, 1)
	rd.compute_list_end()

	# Submit to GPU and wait for sync
	rd.submit()
	rd.sync()

	# Read back the data from the buffer
	var output_bytes := rd.buffer_get_data(buffer)
	var output := output_bytes.to_float32_array()
	print("Input: ", input)
	print("Output: ", output)

# Called every frame. 'delta' is the elapsed time since the previous frame.
func _process(delta):
	pass

```

All of which should be copied and pasted verbatim from the example.

However, on running it, the actual shader doesn't seem to modify the data:

```
Input: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Output: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```

I assume this is a documentation issue. If instead this is a Godot engine bug I can close this one and open an issue there.

**URL to the documentation page:**

https://docs.godotengine.org/en/stable/tutorials/shaders/compute_shaders.html

**Comment by clayjohn** [comment]
CC @bruvzg Is it possible this is caused by a MoltenVK update? I don't know of anything on the renderer side that would result in this type of change in behaviour

**Comment by bruvzg** [comment]
Shader seems to be working on M1 Mac (both 4.2 official build and custom 4.x with the custom MoltenVK).
```
Input: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Output: [2, 4, 6, 8, 10, 12, 14, 16, 18, 20]
```

**Comment by clayjohn** [comment]
> Okay, I'm at a loss. After having loaded it in 4.1.3 and then re-loading it in 4.2, it now runs as expected.

Maybe its a shader caching issue. If so, its going to be very hard to reproduce now.

**Comment by TheCleric** [comment]
I believe I have tracked down the root cause. I was using Godot's built in text editor originally to edit the glsl file. This seems to not trigger something in the editor that the glsl file needs to be reevaluated. So it continues using whatever was there when it first saw it (in my case an empty text file).

However when editing it elsewhere, such as VS Code (or opening it in Godot 4.1) makes the editor realize that a change has occurred and picked up on it. The documentation does clearly state to use an external editor, so in the end that's a me problem. Closing. Thanks for the help.

**Comment by majikayogames** [comment]
Thank you @TheCleric. Was pulling my hair out trying to figure out what I was doing wrong with compute shaders, trying to get something basic set up. Then I conceded to just follow the basic tutorial, which I was surprised to find I still couldn't get working, having copied it line by line from the docs, got the same output as you.

I'm wondering if this would still be a bug? Why can't we use the built in text editor for .glsl files? Would Godot also not recognize normal .txt files as being updates using the internal editor? I ran into another (I think definite bug) today with a very similar pattern relating to resources/scenes not updating and getting hardlocked like this which I'm struggling to reproduce.

When in doubt, rename your .godot folder and see if it magically fixes everything. Has worked for me numerous times.
